wrapper/src/main/java/software/amazon/jdbc/plugin/cache/CacheConnection.java

```diff
@@ -257,10 +328,197 @@ private void returnConnectionBackToPool(StatefulRedisConnection <byte[], byte[]>
     }
   }
 
+  private boolean hasIamAuthentication() {
+    return CACHE_IAM_USERNAME.getString(this.connectionProperties) != null &&
+        !CACHE_IAM_USERNAME.getString(this.connectionProperties).trim().isEmpty();
+  }
+
+  private boolean hasTraditionalAuthentication() {
+    return CACHE_USERNAME.getString(this.connectionProperties) != null &&
+        CACHE_PASSWORD.getString(this.connectionProperties) != null;
+  }
+
+  private void refreshIamTokenIfExpired() {
+    String username = CACHE_IAM_USERNAME.getString(this.connectionProperties);
+    String region = CACHE_IAM_REGION.getString(this.connectionProperties);
+    String hostname = this.hostAndPort[0];
+    int port = this.hostAndPort.length > 1 ? Integer.parseInt(this.hostAndPort[1]) : 6379;
+
+    if (username == null || hostname == null || region == null) {
+      throw new RuntimeException("Missing required IAM authentication properties");
+    }
+
+    String cacheKey = IamAuthUtils.getCacheKey(username, hostname, port, Region.of(region));
+    TokenInfo tokenInfo = IamAuthCacheHolder.getCachedToken(cacheKey);
+    if (tokenInfo == null || tokenInfo.isExpired()) {
+      ElastiCacheIamTokenUtility tokenUtility = new ElastiCacheIamTokenUtility();
+
+      String cacheName = CACHE_NAME.getString(this.connectionProperties);
+      if (cacheName != null) {
+        tokenUtility.setCacheName(cacheName);
+        LOGGER.finest("Using ElastiCache cache name: " + cacheName);
+      }
+      String token = tokenUtility.generateAuthenticationToken(
+          AwsCredentialsManager.getProvider(null, this.connectionProperties),
+          Region.of(region),
+          hostname,
+          port,
+          username);
+
+      connectionProperties.setProperty("iamToken", token);
```

Comment:
Please don't store the IAM token in 2 different places. This will make it likely for programming error keeping them in sync.